### PR TITLE
Latency improvement

### PR DESCRIPTION
**Pull request by jjsjann123**
1. removes some redundant check of `Fusion::hasReduction` on reduction heuristics;
2. caches reduction_tv_ in `FusionExecutorCache` to avoid repetitive lookup at runtime;
3. improves `InputsIdLookup::lookupId` performance by reusing string buffer instead of using stringstream;

### CODE REVIEW

**Review by csarofeen — APPROVED** [review]
LGTM

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -7486,7 +7486,7 @@ TEST(NVFuserTest, FusionInputsIdLookup_CUDA) {
   at::Tensor t2 = at::randn({6, 4}, options);
 
   // create a cache with max size 2;
-  auto inputs_id_lookup = InputsIdLookup(2);
+  torch::jit::fuser::cuda::InputsIdLookup inputs_id_lookup(2);
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp:7489]: what is the reason for this change?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 2]
```diff
@@ -204,35 +204,45 @@ at::DimVector inversePermutation(
   }
 }
 
+template <typename T, typename CONTAINER>
+void encode_buf(T value, CONTAINER& buffer) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: 1. naming: encodeBuff (or even better, `encodeBuffer` since we don't need to save 2 keystrokes here)

2. why does it have to be a template?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 3]
```diff
@@ -204,35 +204,45 @@ at::DimVector inversePermutation(
...
+  char* v = reinterpret_cast<char*>(&value);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: `const char`

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 4]
```diff
@@ -204,35 +204,45 @@ at::DimVector inversePermutation(
...
+  for (int i = 0; i < sizeof(T); i++) {
+    buffer.push_back(*(v++));
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:210]: since we care about perf, it would be preferable to avoid push_back (and resize + indexing) or at least reserve. It's also a reason not to make this over generic and stick to a container type.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 5]
```diff
@@ -204,35 +204,45 @@ at::DimVector inversePermutation(
...
+}
+
 } // namespace
 
 InputsIdLookup::IdLookupReturn InputsIdLookup::lookupId(
     const at::ArrayRef<IValue>& inputs) {
   IdLookupReturn ret;
-  std::stringstream encoded_inputs;
+
+  std::lock_guard<std::mutex> guard(mutex_);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:221]: what is this mutex guarding? we didn't have a mutex here before.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 6]
```diff
@@ -266,61 +275,62 @@ FusionExecutorCache::FusionExecutorCache(std::unique_ptr<Fusion>&& fusion)
...
+
   // get unique id `unique_id` for given input set `inputs`;
   auto id_lookup_ret = inputs_id_lookup_.lookupId(inputs);
   if (id_lookup_ret.eviction) {
     evictCache(id_lookup_ret.evict_id);
   }
 
-  const size_t unique_id = id_lookup_ret.id;
-  const int device_index = getCommonDeviceCUDA(inputs);
+  size_t unique_id = id_lookup_ret.id;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: what is the reason for removing const?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 7]
```diff
@@ -32,6 +33,8 @@ class TORCH_CUDA_API InputsIdLookup {
   explicit InputsIdLookup(size_t max_cache_size = 10)
       : max_cache_size_(max_cache_size){};
 
+  InputsIdLookup(const InputsIdLookup&) = delete;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: use NoCopySemantics from utils.h instead

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 8]
```diff
@@ -52,6 +55,11 @@ class TORCH_CUDA_API InputsIdLookup {
   }
 
  private:
+  // string to store encoded input meta information. Reuse the buffer instead of
+  // stringtream gives few us perf gain.
+  std::string encoding_;
+  std::mutex mutex_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h:61]: Add a comment for the mutex? Why do we need it?

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 9]
```diff
@@ -345,6 +337,7 @@ void scheduleReduction(
     const ReductionParams& rparams,
     TensorView* red_tv,
     std::vector<TensorView*> outs_of_red) {
+  FUSER_PERF_SCOPE("scheduleReduction");
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp:340]: nice

File: test/cpp/jit/test_gpu.cpp (hunk 1 above)
Comment (jjsjann123) [reply test/cpp/jit/test_gpu.cpp:7489]: I added a mutex in the struct so it's not copyable. We don't really need copy so I'm keeping it simple.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: 1. oops. :stuck_out_tongue: 
2. I was playing with different container & types. I can get rid of the template as it's not necessary any more.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 4 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:210]: Ha, this is the part I was playing with earlier.
We do reuse the container object between iteration. So it grows as needed, but push_back shouldn't change the capacity and there shouldn't be any difference between this and (resize + indexing).

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:210]: I think this also applies to other sequential container in std... But I'm not that confident.

I agree that having a CONTAINER template argument here is ambiguous. I'll update it.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 5 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:221]: `encoding_` is now the buffer that was re-used. Back then we are not reusing containers but doing stringstream, getting rid of the copy gives us ~4us overhead.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 6 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: Good catch! I'll revert it.

I was doing measurement in with our tracing earlier and updated the value in a separate block.

**Comment by jjsjann123** [comment]
I think I got all your comments, tagging @tlemo again to get an approval to merge.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 5 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:221]: can you please add a comment linking the mutex with the shared state it guards? (where we define `encoding_` for example, it can be as brief as `// Shared state, guarded by mutex_`

**Review by tlemo — APPROVED** [review]
